crates/red_knot_python_semantic/src/types/signatures.rs

```diff
@@ -105,42 +95,209 @@ impl<'db> Parameters<'db> {
-            .iter()
-            .map(|arg| ParameterWithDefault::from_node(db, definition, arg))
-            .collect();
+            .map(|arg| Parameter::PositionalOnly(ParamWithDefault::from_node(db, definition, arg)));
+        let positional_or_keyword = args.iter().map(|arg| {
+            Parameter::PositionalOrKeyword(ParamWithDefault::from_node(db, definition, arg))
+        });
         let variadic = vararg
             .as_ref()
-            .map(|arg| Parameter::from_node(db, definition, arg));
+            .map(|arg| Parameter::Variadic(Param::from_node(db, definition, arg)));
         let keyword_only = kwonlyargs
             .iter()
-            .map(|arg| ParameterWithDefault::from_node(db, definition, arg))
-            .collect();
+            .map(|arg| Parameter::KeywordOnly(ParamWithDefault::from_node(db, definition, arg)));
         let keywords = kwarg
             .as_ref()
-            .map(|arg| Parameter::from_node(db, definition, arg));
-        Self {
-            positional_only,
-            positional_or_keyword,
-            variadic,
-            keyword_only,
-            keywords,
+            .map(|arg| Parameter::Keywords(Param::from_node(db, definition, arg)));
+        Self(
+            positional_only
+                .chain(positional_or_keyword)
+                .chain(variadic)
+                .chain(keyword_only)
+                .chain(keywords)
+                .collect(),
+        )
+    }
+
+    pub(crate) fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub(crate) fn iter(&self) -> std::slice::Iter<Parameter<'db>> {
+        self.0.iter()
+    }
+
+    /// Iterate initial positional parameters, not including variadic parameter, if any.
+    ///
+    /// For a valid signature, this will be all positional parameters. In an invalid signature,
+    /// there could be non-initial positional parameters; effectively, we just won't consider those
+    /// to be positional, which is fine.
+    pub(crate) fn positional_only(&self) -> impl Iterator<Item = &Parameter<'db>> {
+        self.iter().take_while(|param| param.is_positional())
+    }
+
+    /// Return number of positional parameters this signature can accept.
+    ///
+    /// Doesn't account for variadic parameter.
+    pub(crate) fn positional_count(&self) -> usize {
+        self.positional_only().count()
+    }
+
+    /// Return parameter at given index, or `None` if index is out-of-range.
+    pub(crate) fn get(&self, index: usize) -> Option<&Parameter<'db>> {
+        self.0.get(index)
+    }
+
+    /// Return positional parameter at given index, or `None` if `index` is out of range.
+    ///
+    /// Does not return variadic parameter.
+    pub(crate) fn get_positional(&self, index: usize) -> Option<&Parameter<'db>> {
+        if let Some(candidate) = self.get(index) {
+            if candidate.is_positional() {
+                return Some(candidate);
+            }
         }
+        None
```

Comment:
```suggestion
        self.get(index)
            .and_then(|candidate| candidate.is_positional().then_some(candidate))
```

Reply:
Not totally convinced this is more readable, but it is shorter, so sure :)

Reply:
hehe, I'm not wedded to it!